vnext/Microsoft.ReactNative/Utils/TextHitTestUtils.cpp

```diff
@@ -0,0 +1,176 @@
+  // Y └───────────┘  └───────────┘
+  //        LTR            RTL
+  const auto isBeforeX = !isRtl ? point.X < rect.X : point.X > right;
+  return point.Y < rect.Y || (point.Y < bottom && isBeforeX);
+}
+
+static bool IsRTL(const winrt::TextPointer &textPointer) {
+  auto currentPointer = textPointer.GetPositionAtOffset(1, winrt::LogicalDirection::Forward);
+  auto currentRect = textPointer.GetCharacterRect(winrt::LogicalDirection::Forward);
+  auto firstCharacterRect = currentRect;
+  while (currentPointer != nullptr) {
+    currentRect = currentPointer.GetCharacterRect(winrt::LogicalDirection::Forward);
+    // If we haven't figured out the direction by the time we reach the end of
+    // a line, try again for the next line.
+    if (currentRect.Y != firstCharacterRect.Y) {
+      firstCharacterRect = currentRect;
+    } else if (currentRect.X != firstCharacterRect.X) {
+      return currentRect.X < firstCharacterRect.X;
+    }
+
+    currentPointer = currentPointer.GetPositionAtOffset(1, winrt::LogicalDirection::Forward);
+  }
+
+  // Assume LTR if there are not enough characters to determine LTR vs. RTL.
+  // This may occur for TextBlocks with a single character per line.
+  return false;
+}
+
+static winrt::TextPointer GetPositionFromPointCore(
+    const winrt::TextPointer &start,
+    const winrt::TextPointer &end,
+    const winrt::Point &targetPoint) {
+  // Since characters in a TextBlock are sorted from top-left to bottom-right
+  // (or top-right to bottom-left for RTL), we can use binary search to find
+  // the character with bounds that container the pointer point.
+  //
+  // This algorithm currently makes the following assumptions:
+  // 1. Characters on the same line have the same Rect::Y value
+  // 2. Search space is over only LTR or only RTL characters
+  const auto width = start.VisualParent().Width();
+  const auto isRtl = IsRTL(start);
+  auto textPointer = start;
+  auto L = start.Offset();
+  auto R = end.Offset();
+  while (L <= R) {
+    const auto m = /* floor */ (L + R) / 2;
+    const auto relativeOffset = m - textPointer.Offset();
+    textPointer = textPointer.GetPositionAtOffset(relativeOffset, winrt::LogicalDirection::Forward);
+    const auto rect = textPointer.GetCharacterRect(winrt::LogicalDirection::Forward);
+    if (IsPointAfterCharacter(targetPoint, textPointer, rect, width, isRtl) /* A[m] < T */) {
+      L = m + 1;
+    } else if (IsPointBeforeCharacter(targetPoint, textPointer, rect, width, isRtl) /* A[m] > T */) {
+      R = m - 1;
+    } else {
+      return textPointer;
+    }
+  }
+
+  return nullptr;
+}
+
+winrt::TextPointer TextHitTestUtils::GetPositionFromPoint(const winrt::TextBlock &textBlock, const winrt::Point &targetPoint) {
+  return GetPositionFromPointCore(textBlock.ContentStart(), textBlock.ContentEnd(), targetPoint);
+}
+
+winrt::TextPointer TextHitTestUtils::GetPositionFromPoint(const winrt::Run &run, const winrt::Point &targetPoint) {
+  const auto start = run.ContentStart();
```

Comment:
The bounding box check code was taken from @rectified95's previous approach:
https://github.com/microsoft/react-native-windows/blob/master/vnext/Microsoft.ReactNative/Views/TouchEventHandler.cpp#L471-L492